Bentheburrito/sacastats
# New ESS Health Monitoring + Event Processing Strategy to Increase ESS Robustness

This PR overhauls the way SacaStats processes events from the ESS, and adds new health checks to automatically restart seemingly unhealthy/flaky websocket connections based on other connections. Many new modules are included to accomplish this, particularly `EventTracker.{Deduper, Manager, Report, Supervisor}`. The `EventTracker` module itself is now a `GenServer` with its own `%Report{}` state that it populates as events are consumed. the `Manager` regularly evaluates these reports and tells the `Supervisor` to restart unhealthy connections. Since there are now multiple `PS2.Socket`s running, this PR also introduces a `Deduper` to collapse duplicate events into one instance before attempting to insert into the DB.

Please lmk your questions in the review/discord :)

This will close #107 

**How to Test**

0. You should configure two new environment variables in your `.env`:
```
SERVICE_ID_LIST=<SID1>,<SID2>,[...]
NUM_EVENT_TRACKERS=2
```
I'll send you the actual SIDs privately, of course.
1. When you start the app, you should see multiple copies of the usual "Connected to socket" and "Received subscription acknowledgement" messages in IEx (you'll see NUM_EVENT_TRACKERS copies). 
2. If there are no errors, type `:observer.start` in IEx and click the "Application" tab.
3. Find the "Elixir.SacaStats.EventTracker.Supervisor" node (image below). It should have two linked processes, which themselves have one linked process.
![image](https://user-images.githubusercontent.com/33915719/177075084-5703ab4a-cab4-4c87-84f6-3ef35e0f0d56.png)
4. Double click on one of the Supervisor's linked processes (in the image, this would be either PID<0.651.0> or PID<0.656.0>)
5. Click the State tab. You should see something similar to the image below. 
![image](https://user-images.githubusercontent.com/33915719/177075147-96d02a5a-0d27-400f-a0c5-0ad8914668e3.png)
6. Spam Ctrl + R to refresh the state. You should see the event counts increasing. Every 15 seconds, the event counts should clear (this means the Manager is collecting reports!)
7. Occasionally, you should see a warning in IEx similar to below. This means the Manager thought a connection was unhealthy, so it restarted it.
```
[warn] Terminating #PID<0.13964.0>
Event count for GainExperience failed by 15 with SID: <redacted SID>
```
8. Finally, exit IEx and run `mix test`. They should all pass.
